### ISSUE TITLE
dtest `TimeoutError: 28 Jun 2023 13:06:56 [node4] Missing: ['Done with off-strategy compaction for ks.cf']`

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#14437]
`compaction_test.TestCompaction.test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]` failed in next promotion

```
            logger.debug("Running offstrategy compaction on node4")
            run_rest_api(node4, f"/storage_service/keyspace_offstrategy_compaction/{ks}?cf={cf}")
>           node4.watch_log_for(f"Done with off-strategy compaction for {self.FULL_TABLE_NAME}", from_mark=log_mark)
```

The waiting for log message timed out, the message never appeared.

I looked into node and test logs.
Node indeed saw the request:
```
INFO  2023-06-28 12:56:52,104 [shard 1] api - perform_keyspace_offstrategy_compaction: keyspace=ks tables={table{name=cf, id=18309370-15b3-11ee-993e-67b6b00fd3b1}}
```

In the test logs I noticed:
```
12:56:52,104 782     urllib3.connectionpool         DEBUG    connectionpool.py   :456  | test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]: http://127.0.24.4:10000 "POST /storage_service/keyspace_offstrategy_compaction/ks?cf=cf HTTP/1.1" 200 5
12:56:52,104 782     tools.cluster                  DEBUG    cluster.py          :76   | test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]: API result: False
```

the API result was False... I checked what this means, it looks like the result of the API call indicates whether compaction was needed in the first place
```cpp
        bool res = false;
...
        auto task = co_await compaction_module.make_and_start_task<offstrategy_keyspace_compaction_task_impl>({}, std::move(keyspace), ctx.db, table_infos, res);
...
        co_return json::json_return_type(res);
```

```cpp
class offstrategy_keyspace_compaction_task_impl : public offstrategy_compaction_task_impl {
private:
    sharded<replica::database>& _db;
    std::vector<table_info> _table_infos;
    bool& _needed;
public:
    offstrategy_keyspace_compaction_task_impl(tasks::task_manager::module_ptr module,
            std::string keyspace,
            sharded<replica::database>& db,
            std::vector<table_info> table_infos,
            bool& needed) noexcept
```

```cpp
future<> offstrategy_keyspace_compaction_task_impl::run() {
    _needed = co_await _db.map_reduce0([&] (replica::database& db) -> future<bool> {
        bool needed = false;
        tasks::task_info parent_info{_status.id, _status.shard};
        auto& module = db.get_compaction_manager().get_task_manager_module();
        auto task = co_await module.make_and_start_task<shard_offstrategy_keyspace_compaction_task_impl>(parent_info, _status.keyspace, _status.id, db, _table_infos, needed);
        co_await task->done();
        co_return needed;
    }, false, std::plus<bool>());
}
```

digging deeper a few levels, `!needed` means that the compaction wasn't performed at all.
```cpp
future<bool> table::perform_offstrategy_compaction() {
    // If the user calls trigger_offstrategy_compaction() to trigger
    // off-strategy explicitly, cancel the timeout based automatic trigger.
    _off_strategy_trigger.cancel();
    bool performed = false;
    co_await parallel_foreach_compaction_group([this, &performed] (compaction_group& cg) -> future<> {
        performed |= co_await _compaction_manager.perform_offstrategy(cg.as_table_state());
    });
    co_return performed;
}
```

In other words, in this case the compaction wasn't performed because it wasn't needed - and that's why the logs didn't show `Done with off-strategy compaction for ks.cf`.

However there remains the question: is Scylla wrong here not to perform the compaction, or is the dtest wrong and it should expect that sometimes compaction won't be needed?

cc @raphaelsc @bhalevy

**Comment by bhalevy** [comment]
The following stats show that there was nothing to repair.
Right @asias ?
```
INFO  2023-06-28 12:56:50,783 [shard 0] repair - repair[b701086c-1805-4fc7-bb65-30bd26ec375e]: stats: repair_reason=repair, keyspace=ks, tables={cf}, ranges_nr=256, round_nr=257, round_nr_fast_path_already_synced=257, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=0, rpc_call_nr=2562, tx_hashes_nr=0, rx_hashes_nr=0, duration=0.11023173 seconds, tx_row_nr=0, rx_row_nr=0, tx_row_bytes=0, rx_row_bytes=0, row_from_disk_bytes={{127.0.24.1, 140}, {127.0.24.2, 0}, {127.0.24.3, 140}, {127.0.24.4, 140}}, row_from_disk_nr={{127.0.24.1, 1}, {127.0.24.2, 0}, {127.0.24.3, 1}, {127.0.24.4, 1}}, row_from_disk_bytes_per_sec={{127.0.24.1, 0.00121122}, {127.0.24.2, 0}, {127.0.24.3, 0.00121122}, {127.0.24.4, 0.00121122}} MiB/s, row_from_disk_rows_per_sec={{127.0.24.1, 9.0718}, {127.0.24.2, 0}, {127.0.24.3, 9.0718}, {127.0.24.4, 9.0718}} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={}
INFO  2023-06-28 12:56:50,783 [shard 0] repair - repair[b701086c-1805-4fc7-bb65-30bd26ec375e]: completed successfully, keyspace=ks
INFO  2023-06-28 12:56:50,842 [shard 1] repair - repair[b701086c-1805-4fc7-bb65-30bd26ec375e]: stats: repair_reason=repair, keyspace=ks, tables={cf}, ranges_nr=256, round_nr=258, round_nr_fast_path_already_synced=258, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=0, rpc_call_nr=2564, tx_hashes_nr=0, rx_hashes_nr=0, duration=0.16901053 seconds, tx_row_nr=0, rx_row_nr=0, tx_row_bytes=0, rx_row_bytes=0, row_from_disk_bytes={{127.0.24.1, 280}, {127.0.24.2, 280}, {127.0.24.3, 0}, {127.0.24.4, 280}}, row_from_disk_nr={{127.0.24.1, 2}, {127.0.24.2, 2}, {127.0.24.3, 0}, {127.0.24.4, 2}}, row_from_disk_bytes_per_sec={{127.0.24.1, 0.00157995}, {127.0.24.2, 0.00157995}, {127.0.24.3, 0}, {127.0.24.4, 0.00157995}} MiB/s, row_from_disk_rows_per_sec={{127.0.24.1, 11.8336}, {127.0.24.2, 11.8336}, {127.0.24.3, 0}, {127.0.24.4, 11.8336}} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={}
INFO  2023-06-28 12:56:50,842 [shard 1] repair - repair[b701086c-1805-4fc7-bb65-30bd26ec375e]: completed successfully, keyspace=ks
```

**Comment by bhalevy** [comment]
It seems like the problem could be in the test, that expects off-strategy compaction after repair, but in this case, repair was a noop so offstrategy didn't happen (which is expected in this case)

**Comment by raphaelsc** [comment]
> It seems like the problem could be in the test, that expects off-strategy compaction after repair, but in this case, repair was a noop so offstrategy didn't happen (which is expected in this case)

Yes, it seems to be a test issue, we need to adjust the dtest to comply with the rules.
